unified_planning/grpc/server.py

```diff
@@ -0,0 +1,165 @@
+
+    def __new__(cls, **kwargs) -> "GRPCPlanner":
+        """Create a thread-safe singleton instance of the planner
+
+        Modes:
+         - If parameters are provided,
+                - If the port is available, create a new client
+                - If the port is already in use, return the existing client
+         - If no parameters are provided, use default.
+        """
+        port = kwargs.get("port", None)
+
+        if (port, cls) not in cls._instances:
+            with cls._lock:
+                if cls not in cls._instances:
+                    cls._ports.add(port)
+                    cls._instances[(port, cls)] = super().__new__(cls)
+                    return cls._instances[(port, cls)]
+        elif kwargs.get("override", False):
+            return super().__new__(cls)
+        return cls._instances[(port, cls)]
+
+    def __del__(self) -> None:
+        """Delete the planner instance"""
+        self._channel.close()
+
+        for instance in self._instances:
+            if instance[1] == self:
+                self._instances.pop(instance)
+                break
+
+    def _solve(
+        self,
+        problem: "up.model.AbstractProblem",
+        heuristic: Optional[
+            Callable[["up.model.state.ROState"], Optional[float]]
+        ] = None,
+        timeout: Optional[float] = None,
+        output_stream: Optional[IO[str]] = None,
+    ) -> "up.engines.results.PlanGenerationResult":
+        """GRPC Client for Unified Planning
+
+        :param problem: Problem to solve
+        :type problem: up.model.AbstractProblem
+        :param heuristic:  is a function that given a state returns its heuristic value or `None` if the state is a dead-end, defaults to `None`.
+        :type heuristic: Optional[Callable[["up.model.state.ROState"], Optional[float]], None], Optional
+        :param timeout: Timeout in seconds, defaults to None
+        :type timeout: Optional[float], optional
+        :param output_stream: Log output stream, defaults to None
+        :type output_stream: Optional[IO[str]], optional
+        :return: Plan generation result
+        :rtype: up.engines.results.PlanGenerationResult
+        """
+
+        # Assert that the problem is a valid problem
+        assert isinstance(problem, up.model.Problem)
+
+        proto_problem = self._writer.convert(problem)
+
+        req = proto.PlanRequest(problem=proto_problem, timeout=timeout)
+        response_stream = self._planner.planOneShot(req)
+        response = self._reader.convert(response_stream, problem)
+        assert isinstance(response, up.engines.results.PlanGenerationResult)
+        if (
+            response.status == PlanGenerationResultStatus.INTERMEDIATE
+            and heuristic is not None
+        ):
+            # TODO: Implement heuristic
+            pass
+        else:
+            return response
+
+        raise UPException("No response from the server")
```

Comment:
The `INTERMEDIATE` result status is not a valid status for a oneshot planning call. It is used only by the anytime planning operation mode.

I would change the code as follow:
```suggestion
        if heuristic is not None:
            warn("Custom heuristic is not used by this planner!")
        return response
```

Reply:
Thanks!